### ISSUE TITLE
`onComplete` is invoked after rendering with loaded data in 3.8.x

### ISSUE THREAD

**Issue by dfilatov** [apollographql/apollo-client#11209]
### Issue Description

Something has been broken in 3.8.x versions. `onCompleted` now is invoked after rendering with loaded data, it's unexpected and leads to errors in our application because we need to do some additional work with loaded data before it's rendered. In 3.7.17 version this works as expected.

### Link to Reproduction

https://codesandbox.io/s/sad-fermat-djg7mq

### Reproduction Steps

  * open https://codesandbox.io/s/sad-fermat-djg7mq
  * take a look at console, you'll see:
      `render true`
      `completed`
      `render false`
  * change version of `@apollo/client` to `3.8.3`, save
  * take a look at console, you'll see:
      `render true`
      `render false`
      `completed`

**Comment by phryneas** [comment]
3.8 includes a change that is generally recommended by the React team when dealing with async data sources (like Apollo Client): It calls the `handleStoreChange` callback of `useSyncExternalStore` instead of a `useState` `setState` callback like before.
As a result, React not applies different timing of things here.

Generally, please note that we don't have a lot of control over the order of operation here - React decides pretty arbitrarily when it rerenders. 
We actually have always called the code triggering `onCompleted` after we trigger the rerender. Until now, it caused the React state update batching behavior to move the next render back a tick. `handleStoreChange` should also do this (to batch with other external state source updates), but sometimes it doesn't.
Keep in mind, this is "autobatching" is only the case in React 18 - in React 17 you would see the current behaviour even with older versions of Apollo Client.

Once React [fixes this bug](https://github.com/facebook/react/issues/25191) the timing will likely go back to the behavior you have seen in React 18 with 3.7.

=> We have never been able to guarantee a certain order of operation, and can also not do so going into the future. React decides when it actually rerenders, not us.


Out of curiosity - what are you doing here that needs to happen before a render?

**Comment by dfilatov** [comment]
@phryneas Thank you for quick response!

> Out of curiosity - what are you doing here that needs to happen before a render?

We are building/committing/uncommitting/resetting our internal models based on query data within `onCompleted`/`onError`

**Comment by phryneas** [comment]
Hm, if that is local component state, most of the time you should be able to do that synchronously during component render - can you share some example code here? Maybe I can suggest an alternative way of doing so.

**Comment by dfilatov** [comment]
Very simplified approach looks like:
```tsx
const MyComponent = ({ model }) => {
    const { loading } = useSomeDataQuery({
        onCompleted(data) {
            model.setData(data);
            model.commit();
        }
    });

    return loading ?
        <Loader/> :
        <ModelView={ model }/>;        
};
```

I could get the same behaviour with `React.useMemo` but that looks very weird and unnatural to me:
```tsx
const MyComponent = ({ model }) => {
    const { data, loading } = useSomeDataQuery({});
    const modelWithData = React.useMemo(
        () => {
            if(loading) {
                return null;
            }

            model.setData(data);
            model.commit();

            return model;
        },
        [loading, data]
    );

    return loading ?
        <Loader/> :
        <ModelView={ modelWithData }/>;        
};
```

**Comment by iamnnort** [comment]
The same issue

**Comment by Mofo50C** [comment]
Yeah, im having this issue as well. Apollo client v3.8.8, react v18.2, nextjs v14. After a mutation, refetchQueries runs, query is updated, data is returned but onCompleted isn't called so the component state doesn't update. Previously working on 3.7

**Comment by jmclain20** [comment]
This is also occurring for us.

In my case even just adding a console.log within the onCompleted it is never called when a refetch or poll happens. Adding notifyOnNetworkStatusChange will allow the onCompleted to fire but this was not the case prior to 3.8 client version. Seeing if moving forward this is needed always or if this will be adjusted.

**Comment by ricardopieper** [comment]
Same

**Comment by phryneas** [comment]
Hi everyone, I'm doing some housekeeping on old issue.

With Apollo Client 4.0, we have removed the `onCompleted` and `onError` callbacks from the `useQuery` and `useLazyQuery` hooks.

We are discussing the reasoning behind the callback deprecation and removal in this issue: [Heads up: Deprecation of useQuery and useLazyQuery lifecycle hooks #12352](https://github.com/apollographql/apollo-client/issues/12352).

For details on the Apollo Client 4.0 release, please see [this announcement blog post](https://www.apollographql.com/blog/announcing-apollo-client-4-0), [the migration guide](https://www.apollographql.com/docs/react/migrating/apollo-client-4-migration) and the [changelog](https://github.com/apollographql/apollo-client/releases/tag/%40apollo%2Fclient%404.0.0).

As a result I believe this issue is no longer actionable can be closed. Please don't hesitate to open new issues if you encounter them.

Have a great day everyone!
